src/materialized/src/bin/materialized/tracing.rs

```diff
@@ -27,8 +31,85 @@ use mz_ore::metrics::{MetricsRegistry, ThirdPartyMetric};
 
 use crate::Args;
 
+fn create_h2_alpn_https_connector() -> ProxyConnector<HttpsConnector<HttpConnector>> {
+    // This accomplishes the same thing as the default
+    // + adding a `request_alpn`
+    let mut http = HttpConnector::new();
+    http.enforce_http(false);
+
+    mz_http_proxy::hyper::connector(HttpsConnector::from((
+        http,
+        tokio_native_tls::TlsConnector::from(
+            native_tls::TlsConnector::builder()
+                .request_alpns(&["h2"])
+                .build()
+                .unwrap(),
+        ),
+    )))
+}
+
+fn configure_opentelemetry_and_init<
+    L: tracing_subscriber::layer::Layer<S> + Send + Sync + 'static,
+    S: ::tracing::Subscriber + Send + Sync + 'static,
+>(
+    stack: tracing_subscriber::layer::Layered<L, S>,
+    runtime: &tokio::runtime::Runtime,
+    opentelemetry_endpoint: Option<String>,
+    opentelemetry_headers: Option<String>,
+) -> Result<(), anyhow::Error>
+where
+    tracing_subscriber::layer::Layered<L, S>: tracing_subscriber::util::SubscriberInitExt,
+    for<'ls> S: tracing_subscriber::registry::LookupSpan<'ls>,
+{
+    if let Some(endpoint) = opentelemetry_endpoint {
+        // Setting up opentel in the background requires we are in a tokio-runtime
+        // context
+        let _guard = runtime.enter();
+
+        // Manually setup an openssl-backed, h2, proxied `Channel`
+        let endpoint = tonic::transport::Endpoint::from_shared(endpoint)?;
+        let channel = endpoint.connect_with_connector_lazy(create_h2_alpn_https_connector())?;
+        let mut mmap = tonic::metadata::MetadataMap::new();
+        // clap guarantees this is set if `opentelemetry_endpoint` is set
+        for header in opentelemetry_headers.unwrap().split(',') {
+            let mut splits = header.splitn(2, '=');
+            let k = splits
+                .next()
+                .context("opentelemetry-headers must be of the form key=value")?;
+            let v = splits
+                .next()
+                .context("opentelemetry-headers must be of the form key=value")?;
+
+            mmap.insert(tonic::metadata::MetadataKey::from_str(k)?, v.parse()?);
+        }
+
+        let otlp_exporter = opentelemetry_otlp::new_exporter()
+            .tonic()
+            .with_channel(channel)
+            .with_metadata(mmap);
+
+        let tracer = opentelemetry_otlp::new_pipeline()
+            .tracing()
+            .with_exporter(otlp_exporter)
+            .install_batch(opentelemetry::runtime::Tokio)
+            .unwrap();
+        let otel_layer = tracing_opentelemetry::layer().with_tracer(tracer);
+
+        let stack = stack.with(otel_layer);
+
+        stack.init();
+        Ok(())
+    } else {
+        stack.init();
+        Ok(())
+    }
+}
```

Comment:
Can I get ya to apply the following style nits? Generally try to `use` things that aren't horribly ambiguous, and keep bounds on generic parameters in a `where` clause unless they're dead simple:

```diff
diff --git a/src/materialized/src/bin/materialized/tracing.rs b/src/materialized/src/bin/materialized/tracing.rs
index 5dca4ff94..581ae5095 100644
--- a/src/materialized/src/bin/materialized/tracing.rs
+++ b/src/materialized/src/bin/materialized/tracing.rs
@@ -19,10 +19,12 @@ use hyper_proxy::ProxyConnector;
 use hyper_tls::HttpsConnector;
 use prometheus::IntCounterVec;
 use tokio::runtime::Runtime;
+use tonic::metadata::{MetadataKey, MetadataMap};
+use tonic::transport::Endpoint;
 use tracing::{Event, Subscriber};
 use tracing_subscriber::filter::{LevelFilter, Targets};
 use tracing_subscriber::fmt;
-use tracing_subscriber::layer::{Context, Layer, SubscriberExt};
+use tracing_subscriber::layer::{Context, Layer, Layered, SubscriberExt};
 use tracing_subscriber::registry::LookupSpan;
 use tracing_subscriber::util::SubscriberInitExt;
 
@@ -48,18 +50,17 @@ fn create_h2_alpn_https_connector() -> ProxyConnector<HttpsConnector<HttpConnect
     )))
 }
 
-fn configure_opentelemetry_and_init<
-    L: tracing_subscriber::layer::Layer<S> + Send + Sync + 'static,
-    S: ::tracing::Subscriber + Send + Sync + 'static,
->(
-    stack: tracing_subscriber::layer::Layered<L, S>,
+fn configure_opentelemetry_and_init<L, S>(
+    stack: Layered<L, S>,
     runtime: &tokio::runtime::Runtime,
     opentelemetry_endpoint: Option<String>,
     opentelemetry_headers: Option<String>,
 ) -> Result<(), anyhow::Error>
 where
-    tracing_subscriber::layer::Layered<L, S>: tracing_subscriber::util::SubscriberInitExt,
-    for<'ls> S: tracing_subscriber::registry::LookupSpan<'ls>,
+    L: Layer<S> + Send + Sync + 'static,
+    S: Subscriber + Send + Sync + 'static,
+    Layered<L, S>: SubscriberInitExt,
+    for<'ls> S: LookupSpan<'ls>,
 {
     if let Some(endpoint) = opentelemetry_endpoint {
         // Setting up opentel in the background requires we are in a tokio-runtime
@@ -67,9 +68,9 @@ where
         let _guard = runtime.enter();
 
         // Manually setup an openssl-backed, h2, proxied `Channel`
-        let endpoint = tonic::transport::Endpoint::from_shared(endpoint)?;
+        let endpoint = Endpoint::from_shared(endpoint)?;
         let channel = endpoint.connect_with_connector_lazy(create_h2_alpn_https_connector())?;
-        let mut mmap = tonic::metadata::MetadataMap::new();
+        let mut mmap = MetadataMap::new();
         // clap guarantees this is set if `opentelemetry_endpoint` is set
         for header in opentelemetry_headers.unwrap().split(',') {
             let mut splits = header.splitn(2, '=');
@@ -80,7 +81,7 @@ where
                 .next()
                 .context("opentelemetry-headers must be of the form key=value")?;
 
-            mmap.insert(tonic::metadata::MetadataKey::from_str(k)?, v.parse()?);
+            mmap.insert(MetadataKey::from_str(k)?, v.parse()?);
         }
 
         let otlp_exporter = opentelemetry_otlp::new_exporter()
```

Reply:
I differ in opinion here typically, but I can see how this cleans things up here!